src/ir/term/mod.rs

```diff
@@ -1414,6 +1536,240 @@ impl Value {
     }
 }
 
+/// replace hashmap lookup with array lookup to speed up
+pub fn eval_value_efficient(
+    term_idx: usize,
+    term_arr: &[NumTerm],
+    val_arr: &mut [Option<Value>],
+    h: &FxHashMap<String, Value>,
+) {
+    let c = &term_arr[term_idx];
+    let v: Value = match &c.op {
+        NumOp::Var(idx) => val_arr[*idx].clone().unwrap(),
+        NumOp::Op(op) => {
+            match op {
+                Op::Var(n, _) => h
```

Comment:
Might it be better to differentiate these two? The old `eval` is more self-contained, such that one can evaluate a term quickly. The new `eval` depends on preprocessing, and its use case would be when one wants to evaluate all terms of a Computation and evaluate them multiple times to amortize the preprocessing cost. So that's two very different use cases.
Moreover, the old `eval` has a lot of references, changing the function would require us to change all of them.

Reply:
I have deduplicated the two copies with a enum `EvalMode`, which takes different values in the two cases, but provide the same interface to the eval logic